# Week1 : DataSet + Model 구성하기


## 1. PyKrx 파헤치기
우선 pykrx로 수집 가능한 데이터 목록을 확인하였다.


In [2]:
from pykrx import stock

In [3]:
## ticker 조회
tickers = stock.get_market_ticker_list()
print (len(tickers))
print (tickers[0:10])

951
['095570', '006840', '027410', '282330', '138930', '001460', '001465', '001040', '079160', '00104K']


KOSPI 시장에 등록된 951개 종목의 Ticker를 모두 확인할 수 있다.

In [4]:
for ticker in tickers[0:10]:
    print (ticker, stock.get_market_ticker_name(ticker))

095570 AJ네트웍스
006840 AK홀딩스
027410 BGF
282330 BGF리테일
138930 BNK금융지주
001460 BYC
001465 BYC우
001040 CJ
079160 CJ CGV
00104K CJ4우(전환)


Ticker가 무슨 순서로 반환되나 궁금했는데, 종목명 Alphabetical order로 반환되는 것을 알 수 있었다.

OHLCV: Open, High, Low, Close, Volume
주식의 시가, 고가, 저가, 종가, 거래량이라는 주요 정보를 확인 가능하다.

In [5]:
df = stock.get_market_ohlcv("20230922")
print(df.head())

            시가      고가      저가      종가     거래량        거래대금   등락률
티커                                                              
095570    4210    4260    4110    4215  241580  1007180610 -0.12
006840   18400   18400   18200   18330    2767    50561890  0.38
027410    3570    3600    3560    3575   40129   143286585 -0.42
282330  143000  145000  141200  143000   48466  6906980100 -0.83
138930    6950    7050    6930    7040  682325  4770776990  0.86


In [6]:
df = stock.get_market_price_change("20230901", "20230922")
print(df.head())

            종목명      시가      종가    변동폭   등락률       거래량         거래대금
티커                                                                 
095570   AJ네트웍스    4140    4215     75  1.81  12129705  53797258250
006840    AK홀딩스   19330   18330  -1000 -5.17    102952   1925209470
027410      BGF    3625    3575    -50 -1.38    623725   2255155755
282330   BGF리테일  156400  143000 -13400 -8.57    504545  76149906000
138930  BNK금융지주    6820    7040    220  3.23  13172581  92161202050


Div/Bps/Per/Eps 등 세부 정보도 확인 가능하다.

In [10]:
df = stock.get_market_fundamental("20230922")
print(df.head())

          BPS    PER   PBR    EPS   DIV   DPS
티커                                           
095570   8076  20.97  0.52    201  6.41   270
006840  41948   0.00  0.44      0  1.09   200
027410  16528  14.47  0.22    247  3.08   110
282330  55724  12.76  2.57  11203  2.87  4100
138930  30468   2.93  0.23   2404  8.88   625


거래량, 거래 대금, 투자자별 거래 실적, 외국인 보유량 등의 정보도 확인 가능하다.
시장 전체/ 티커별/ 특정 종목 에 대해 호출해 정보를 확인할 수 있다.

In [14]:
df = stock.get_market_trading_value_by_date("20230901", "20230922", "005930")
print(df.head())

                    기관합계        기타법인            개인         외국인합계  전체
날짜                                                                  
2023-09-01  139427490300 -1161296400 -690605779000  552339585100   0
2023-09-04  -68052267400   392177000 -176985447700  244645538100   0
2023-09-05  -74350579000  -111740400  -47203512700  121665832100   0
2023-09-06  -59255465000   -53886800   -2481822300   61791174100   0
2023-09-07  -38706263000   375561900  -53537474400   91868175500   0


In [13]:
df = stock.get_market_trading_volume_by_investor("20230901", "20230922", "005930") # 삼성전자
print(df.head())

             매도        매수      순매수
투자자구분                             
금융투자   17496001  22483495  4987494
보험      1576714    930898  -645816
투신      2185597   4328441  2142844
사모      3345234   4904165  1558931
은행        74135    195415   121280


In [16]:
df = stock.get_exhaustion_rates_of_foreign_investment('20230922')
print(df.head())

           상장주식수      보유수량        지분율      한도수량      한도소진률
티커                                                        
000020  27931470   1317219   4.718750  27931470   4.718750
000040  96138465  44258009  46.031250  96138465  46.031250
000050  27415270    296866   1.080078  27415270   1.080078
000070   8564271    705161   8.226562   8564271   8.226562
000075    304058      4017   1.320312    304058   1.320312


순매수 상위 종목을 확인할 수 있다.
특정 종목이 아닌 전체 시장에 대해 주제를 잡아보고 싶어 유용할 것 같다.

In [17]:
df = stock.get_market_net_purchases_of_equities("20230901", "20230922", "KOSPI", "개인")
print(df.head())

             종목명     매도거래량     매수거래량   순매수거래량         매도거래대금         매수거래대금  \
티커                                                                            
005490  POSCO홀딩스  10022415  10966828   944413  5760590868000  6308538012000   
373220  LG에너지솔루션    975566   1586967   611401   498478974000   810010465500   
051910      LG화학    897826   1311951   414125   502150126000   723282329000   
003670    포스코퓨처엠   4169795   4662972   493177  1734495216500  1938789703500   
096770   SK이노베이션   2944169   4104267  1160098   491555149000   681441494700   

             순매수거래대금  
티커                    
005490  547947144000  
373220  311531491500  
051910  221132203000  
003670  204294487000  
096770  189886345700  


계속해서, 전체 시장에 대한 정보 확인에 도움이 될 인덱스 조회 API도 있었다.

In [18]:
tickers = stock.get_index_ticker_list()
for ticker in stock.get_index_ticker_list():
    print(ticker, stock.get_index_ticker_name(ticker))

1001 코스피
1002 코스피 대형주
1003 코스피 중형주
1004 코스피 소형주
1005 음식료품
1006 섬유의복
1007 종이목재
1008 화학
1009 의약품
1010 비금속광물
1011 철강금속
1012 기계
1013 전기전자
1014 의료정밀
1015 운수장비
1016 유통업
1017 전기가스업
1018 건설업
1019 운수창고업
1020 통신업
1021 금융업
1024 증권
1025 보험
1026 서비스업
1027 제조업
1028 코스피 200
1034 코스피 100
1035 코스피 50
1150 코스피 200 커뮤니케이션서비스
1151 코스피 200 건설
1152 코스피 200 중공업
1153 코스피 200 철강/소재
1154 코스피 200 에너지/화학
1155 코스피 200 정보기술
1156 코스피 200 금융
1157 코스피 200 생활소비재
1158 코스피 200 경기소비재
1159 코스피 200 산업재
1160 코스피 200 헬스케어
1167 코스피 200 중소형주
1182 코스피 200 초대형제외 지수
1224 코스피 200 비중상한 30%
1227 코스피 200 비중상한 25%
1232 코스피 200 비중상한 20%
1244 코스피200제외 코스피지수
1894 코스피 200 TOP 10


인덱스의 OHLCV, 등락률, Fundamental (PER/PBR/배당수익률) 등의 정보를 불러올 수 있다.

In [19]:
df = stock.get_index_ohlcv("20230901", "20230922", "1001")
print(df.head())

코스피              시가       고가       저가       종가        거래량            거래대금  \
날짜                                                                          
2023-09-01  2551.93  2568.42  2551.04  2563.71  493367317  10029746225521   
2023-09-04  2566.93  2584.55  2560.32  2584.55  342072516   8699609485382   
2023-09-05  2578.91  2586.65  2575.19  2582.18  384579023   7590068827455   
2023-09-06  2577.41  2585.98  2560.09  2563.34  422126144   7626693242222   
2023-09-07  2554.87  2562.42  2539.24  2548.26  377270533   8325826544902   

코스피                   상장시가총액  
날짜                            
2023-09-01  2033321986422776  
2023-09-04  2050034095713728  
2023-09-05  2048141824784019  
2023-09-06  2033216059472071  
2023-09-07  2021630331132979  


In [21]:
df = stock.get_index_price_change("20230901", "20230922", "KOSPI")
print(df.head())

                   시가       종가       등락률         거래량             거래대금
지수명                                                                  
코스피           2556.27  2508.13 -1.879883  6579159795  137139862796418
코스피 200        334.75   332.02 -0.819824  1638567429   96473847463513
코스피 100       2520.95  2500.72 -0.799805  1079781716   80290572961747
코스피 50        2309.65  2289.67 -0.870117   728073149   65753024154491
코스피 200 중소형주  1209.82  1188.67 -1.750000   558785713   16183274501766


In [22]:
df = stock.get_index_fundamental("20230901", "20230922", "1001")
print(df.head())

                 종가   등락률    PER  선행PER   PBR  배당수익률
날짜                                                  
2023-09-01  2563.71  0.29  15.68    0.0  0.94   1.93
2023-09-04  2584.55  0.81  15.81    0.0  0.95   1.92
2023-09-05  2582.18 -0.09  15.80    0.0  0.95   1.92
2023-09-06  2563.34 -0.73  15.68    0.0  0.94   1.93
2023-09-07  2548.26 -0.59  15.59    0.0  0.93   1.94


공매도/ ETX / 채권 등에 대한 API도 있지만 이번 프로젝트에서는 사용하지 않을 것 같아 제외하였다.

## 2. DataSet 구성하기

특정 종목에 대해 집중하기 보다는, 전체적인 시장 흐름을 예측하는 모델을 해보고 싶었다. 과거 시간대 수치의 변화를 가지고 예측하는 것은 경제학적 인사이트가 부족해 이보다는, 실제 사회적인 분위기와 연관지어 예측하는 모델을 만들어 보면 좋을 것 같았고, 뉴스 헤드라인을 크롤링해 이를 바탕으로 코스피 지수의 등락을 예측하는 자연어 처리 모델을 설계해보기로 하였다.

### 2.1 News Headline Crawling

뉴스 헤드라인 데이터는 네이버 뉴스에서 크롤링하기로 하였다. 사이트를 분석해 본 결과 2021년 1월 1일부터 현재까지 각 날짜별로 언론사 인기 헤드라인을 확인할 수 있었다. 또, 각 날짜별로 랜덤하게 12개 언론사의 Top 5 헤드라인이 표시된다. 이 방법은 언론사 별로 겹치는 헤드라인이 많아, 한 언론사에서 모든 헤드라인을 추출하기로 하였다. 이 경우 2020년 11월 16일부터 언론사의 20개 헤드라인을 확인할 수 있었다. 특별한 의미는 없고, 언론사 코드 001을 사용하는 연합뉴스의 헤드라인을 크롤링하기로 하였다.

In [36]:
## Import modules
import requests
from bs4 import BeautifulSoup
import datetime
BASE_URL = 'https://media.naver.com/press/001/ranking?type=popular&date='

In [46]:
def CrawlHeadlines(date):
    ## Get HTML
    response = requests.get(BASE_URL + date)
    html = response.text

    ## parse class "press_ranking_list"
    soup = BeautifulSoup(html, 'html.parser')
    ranking_list = soup.select('.list_title')
    headlines = []
    ## parse all text in ranking_list
    for headline in ranking_list:
        headlines.append(headline.text)
    return headlines


Parsing이 제대로 일어나는 것을 확인하기 위해 오늘 날짜의 헤드라인을 확인해보았다.

In [47]:
now = datetime.datetime.now()
nowDate = now.strftime('%Y%m%d')
print(CrawlHeadlines(nowDate))

['"지하철 타는 게 낙…" NYT \'무료승차\' 韓노인의 하루 조명', "한총리에 '뼈있는 발언' 시진핑…中 발표문엔 '방한' 언급 없어", '[아시안게임] \'올림픽 이어 기수\' 구본길 "김서영과 함께라 부담 덜 해"', "서울 송파·김포서 일가족 5명 숨져…'돈문제 갈등' 유서(종합2보)", '친환경·디지털·스마트…제19회 항저우 아시안게임 화려한 개막(종합)', '흑해함대 공격한 우크라 "러 고위 지휘관 숨져…수십명 사상"', '[아시안게임] 북한 인공기 게양 금지인데…개회식서도 펄럭(종합)', '尹대통령 "천안·홍성 국가첨단산업단지 조성 빈틈없이 챙길 것"', '시진핑, 한총리에 "방한 진지 검토…한반도 평화 안정에 노력"(종합)', '9개월 전과 다르네?… 미국 찾아간 젤렌스키에 미묘한 온도차', '英서 140여년 전 숨진 에티오피아 왕자, 머리카락만 고국 품으로', '친환경·디지털·스마트…제19회 항저우 아시안게임 화려한 개막', '이재명, 의료진 권고에 24일차에 단식 중단…"법원 출석"(종합2보)', '이재명 체포동의안 가결에 비명계 의원들 살인예고…40대 검거', '14일만에 달 남극에 해가 떴지만…영원히 잠든 인도 달착륙선', '"반갑다 인왕봉" 57년만에 빗장푼 무등산 정상…尹대통령도 축전(종합)', "뉴욕증시, 연준 당국자들 '매파 발언'에 하락…다우 0.31%↓ 마감", '[아시안게임] 북한 인공기 게양 금지인데…곳곳에서 버젓이 펄럭', '디즈니코리아 대표 "\'무빙\' 시즌2 당연히 검토…의지 확고해"', '포스트 말론이 들어 올린 젊음의 축배…"지금 더 사랑하라"']


In [42]:
Columns = ['Date']
for i in range (1, 21):
    Columns.append('Headline' + str(i))

print (Columns)

['Date', 'Headline1', 'Headline2', 'Headline3', 'Headline4', 'Headline5', 'Headline6', 'Headline7', 'Headline8', 'Headline9', 'Headline10', 'Headline11', 'Headline12', 'Headline13', 'Headline14', 'Headline15', 'Headline16', 'Headline17', 'Headline18', 'Headline19', 'Headline20']


In [53]:
## Create a CSV 
import csv
f = open('Headlines.csv', 'w', encoding='utf-8', newline='')
wr = csv.writer(f)
wr.writerow(Columns)

# Loop date from 2020-11-16 to 2023-09-24
start_date = datetime.date(2020, 11, 16)
end_date = datetime.date(2023, 9, 24)
delta = datetime.timedelta(days=1)

while start_date <= end_date:
    date = start_date.strftime('%Y%m%d')
    headlines = CrawlHeadlines(date)
    if len(headlines) == 0: ## 일부 날짜 데이터가 누락된 경우가 있음
        start_date += delta
        continue
    headlines.insert(0, date)
    wr.writerow(headlines)
    start_date += delta

f.close()

### 2.2 KOSPI 등락 수집

같은 기간에 대해 KOSPI 지수의 등락 정보도 PyKrx를 이용해 수집하였다. 우선 후에 사용할 수 있으니 등락률을 그대로 저장하기로 하였다.

In [63]:
df = stock.get_index_fundamental("20201116", "20230924", "1001")
df = df['등락률']

# 날짜 양식 변경 (YYYYMMDD)
df.index = df.index.map(lambda x: str(x).replace("-", ""))
df.index = df.index.map(lambda x: str(x)[0:8])
print(df.head())

## Save as CSV
df.to_csv("KOSPI.csv", mode='w')

날짜
20201116    1.97
20201117   -0.15
20201118    0.26
20201119    0.07
20201120    0.24
Name: 등락률, dtype: float64


주의하여야 할 점은 시장이 쉬는 날의 등락률을 0으로 처리해 사용할지, 해당 날짜는 완전히 제외하고 사용할지를 고민해 보아야 한다.

## 3. Model 구성하기